<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#While-Loop" data-toc-modified-id="While-Loop-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>While Loop</a></span><ul class="toc-item"><li><span><a href="#Forever-while-loop" data-toc-modified-id="Forever-while-loop-0.1.1"><span class="toc-item-num">0.1.1&nbsp;&nbsp;</span>Forever while loop</a></span></li><li><span><a href="#Exercises" data-toc-modified-id="Exercises-0.1.2"><span class="toc-item-num">0.1.2&nbsp;&nbsp;</span>Exercises</a></span></li></ul></li></ul></li><li><span><a href="#For-Loop" data-toc-modified-id="For-Loop-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>For Loop</a></span><ul class="toc-item"><li><span><a href="#for-loop-syntax" data-toc-modified-id="for-loop-syntax-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>for loop syntax</a></span></li><li><span><a href="#lists/pandas-series:-for-loop" data-toc-modified-id="lists/pandas-series:-for-loop-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>lists/pandas series: for loop</a></span></li><li><span><a href="#for-loop:-list-with-index" data-toc-modified-id="for-loop:-list-with-index-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>for loop: list with index</a></span></li><li><span><a href="#Loop-over:-strings" data-toc-modified-id="Loop-over:-strings-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Loop over: strings</a></span><ul class="toc-item"><li><span><a href="#Exercises" data-toc-modified-id="Exercises-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Exercises</a></span></li></ul></li><li><span><a href="#list-of-lists:-Loop-over" data-toc-modified-id="list-of-lists:-Loop-over-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>list of lists: Loop over</a></span></li><li><span><a href="#Dictionary:-Loop-Over" data-toc-modified-id="Dictionary:-Loop-Over-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Dictionary: Loop Over</a></span></li><li><span><a href="#Numpy-Array:-Loop-Over" data-toc-modified-id="Numpy-Array:-Loop-Over-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Numpy Array: Loop Over</a></span></li><li><span><a href="#Two-Dimensional-Numpy-Array:-Loop-Over" data-toc-modified-id="Two-Dimensional-Numpy-Array:-Loop-Over-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Two-Dimensional Numpy Array: Loop Over</a></span><ul class="toc-item"><li><span><a href="#Exercises" data-toc-modified-id="Exercises-1.8.1"><span class="toc-item-num">1.8.1&nbsp;&nbsp;</span>Exercises</a></span></li></ul></li><li><span><a href="#Pandas-DataFrame:-Loop-Over" data-toc-modified-id="Pandas-DataFrame:-Loop-Over-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Pandas DataFrame: Loop Over</a></span></li><li><span><a href="#Add-column:-DF" data-toc-modified-id="Add-column:-DF-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Add column: DF</a></span></li><li><span><a href="#apply()-function" data-toc-modified-id="apply()-function-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>apply() function</a></span></li><li><span><a href="#use-of-len-function-on-a-specific-df-column-by-apply-function" data-toc-modified-id="use-of-len-function-on-a-specific-df-column-by-apply-function-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>use of len function on a specific df column by apply function</a></span></li><li><span><a href="#use-of-upper-function-on-a-specific-df-column-by-apply-function" data-toc-modified-id="use-of-upper-function-on-a-specific-df-column-by-apply-function-1.13"><span class="toc-item-num">1.13&nbsp;&nbsp;</span>use of upper function on a specific df column by apply function</a></span><ul class="toc-item"><li><span><a href="#Exercises" data-toc-modified-id="Exercises-1.13.1"><span class="toc-item-num">1.13.1&nbsp;&nbsp;</span>Exercises</a></span></li></ul></li></ul></li></ul></div>

## While Loop
* if-elif-else construct, a way to alter the flow of your scripts
* as soon as python encounters an if statement, it checks the condition. When the condtion is true, the corresponding code is executed. If the condition evaluates to False, and if there are elif and else statement, Python continues his search. Anyways python goes throught this code only once. After that, Python heads over to the next command in the script.
* while loop is less used. But in some cases, it can be very helpful.
* the syntax of the while loop is very similar to the if statement

* It executes the code inside if the condition is true. However, as opposed to the if statement.

* **while loop = repeated if statement**
* **or repeating action untill condition is met**
* **as long as the condition is true, the while loop will continue to execute his code over and over again**
* **while loop stops when the condition is fale**

In [ ]:
while condition:
    expression

* Keep going untill the error is no longer above 1

In [5]:
error = 50 
# While loop executes as long as the error variable is above 1
# if the error variable is greater than 1 (condtion), keep running the code over and over again
while error > 1:
    print(error)
    error = error/4 
    print(error)

50
12.5
12.5
3.125
3.125
0.78125


* First run --> error is 50, condition is true (error >1) , execute the code
* Second run --> error is 12.5, condition is true (error > 1), execute the code
* Third run --> error is 3.125, codition is true (error > 1), execute the code
* Fourth run --> error is 0.78125, condition is false (error < 1), stop while loop code                                        

### Forever while loop
* So be careful, your while loop shall stop some time

In [3]:
error = 50
while error > 1:
    print(error)

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
50
5

KeyboardInterrupt: 

### Exercises

In [4]:
# x is 1 now
x = 1
# As long as the x is less than 4 (condition), keep executing code 
while x < 4:
    print(x)
    x = x + 1

1
2
3


First Run --> x is 1, condition is true (x < 4), execute the code (add one)
Second Run --> x is 2, condition is true (x <4), execute the code (add one with previous number)
Third Run --> x is 3, condition is true (x <4), execute the code (add one with previous number)
Fourth Run --> x is 4, condition is false (x >4), stop the code

* As long as the offset is not equal zero (condition), keep executing the code (reduce offset by 1) 

In [ ]:
# Initialize offset
offset = 8

# Code the while loop
while offset != 0:
    print("correcting...")
    offset = offset - 1
    print(offset)

* when ridths cash_in_hand falls to 50,000 tk from 200,000 tk, stop distributing money. Donation of 10,000 tk will be given to 15 students  

In [4]:
cash = 200000
# as long as the cash is greater than 50000 taka (condition), keep executing code
# execution (reduce cash by 10000 taka)
while cash > 50000:
    cash = cash - 10000
    print(cash)

190000
180000
170000
160000
150000
140000
130000
120000
110000
100000
90000
80000
70000
60000
50000


In [ ]:
# Initialize offset
offset = -6

# Code the while loop
while offset != 0 :
    print("correcting...")
    if offset > 0 :
      offset = offset - 1
    else : 
      offset = offset + 1    
    print(offset)

In [5]:
offset = - 6
# as long as the offset is not equal to zero (condition), keep executing the expression
while offset != 0:
    print("correcting..")
    if offset > 0:
        offset = offset - 1
# since the offset value is less than 0, else expression is executed        
    else:
        offset = offset + 1
    print(offset)

correcting..
-5
correcting..
-4
correcting..
-3
correcting..
-2
correcting..
-1
correcting..
0


# For Loop

## for loop syntax
* **for each variable (var/element), in a sequence (seq/list/df/dictionary), execute the expression**

In [ ]:
for var in seq:
    expression

## lists/pandas series: for loop
* print out each element in a list seperately

In [7]:
fam = [1.73, 1.68, 1.71, 1.89]
print(fam[0])
print(fam[1])
print(fam[2])
print(fam[3])

1.73
1.68
1.71
1.89


* Instead of the above repetitive and rediuos approach, you shall use for loop
* Execute some code for each height in the fam list
* I want to do something (expression) on each element (give any arbirary name) from the list (name of original list)
* Process continues untill all heights in fam list have beem iterated over

In [9]:
for height in fam:
    print(height)

1.73
1.68
1.71
1.89


In [10]:
type(height)

float

## for loop: list with index
* use enumerate()
* this time iterate/loop over each index (element1) and height (element2) from list fam

In [13]:
for index, height in enumerate(fam):
    print('index_value', index, ':', height)

index_value 0 : 1.73
index_value 1 : 1.68
index_value 2 : 1.71
index_value 3 : 1.89


* Converted the floats and other data types into one types string while printing using plus (+)

In [19]:
for index, height in enumerate(fam):
    print('index ' + str(index) + ": " + str(height))

index 0: 1.73
index 1: 1.68
index 2: 1.71
index 3: 1.89


## Loop over: strings
* iterates over every character in a string

In [21]:
for c in 'family':
    print(c.capitalize())

F
A
M
I
L
Y


### Exercises

* Write a for loop that iterates over all elements of the areas list and prints out every element separately

In [22]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop
for i in areas:
    print(i)

11.25
18.0
20.0
10.75
9.5


In [26]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Change for loop to use enumerate() and update print()
for index, area in enumerate(areas):
# when printing all data types into string, convert them to string
    print("room " + str(index) + ": " + str(area))

room 0: 11.25
room 1: 18.0
room 2: 20.0
room 3: 10.75
room 4: 9.5


In [32]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

for index, area in enumerate(areas):
    print("room", index, ":", area)

room 0 : 11.25
room 1 : 18.0
room 2 : 20.0
room 3 : 10.75
room 4 : 9.5


* For non-programmer folks, room 0: 11.25 is strange. Wouldn't it be better if the count started at 1

In [30]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop
for index, area in enumerate(areas) :
    print("room " + str(index+1) + ": " + str(area))

room 1: 11.25
room 2: 18.0
room 3: 20.0
room 4: 10.75
room 5: 9.5


##  list of lists: Loop over

* Write a for loop that goes through each sublist of house and prints out the x is y sqm, where x is the name of the room and y is the area of the room.

In [35]:
# house list of lists
house = [["hallway", 11.25], 
         ["kitchen", 18.0], 
         ["living room", 20.0], 
         ["bedroom", 10.75], 
         ["bathroom", 9.50]]
         
# Build a for loop from scratch
for name, area in house:
    print("the", name, "is", area, "sqm")

the hallway is 11.25 sqm
the kitchen is 18.0 sqm
the living room is 20.0 sqm
the bedroom is 10.75 sqm
the bathroom is 9.5 sqm


## Dictionary: Loop Over
* what about loop over other data structures, such as dictionaries and Numpy array
* In both cases, we can use similar for loop construct, but the way you define the "sequence" over which you are iterating will differ depending in the data structure
* use methods items()
* dictionaries are inherently ordered: the order in which they are iterated over is not fixed
* key, values are arbitrary values
* the first variable gets key and second one the value

In [37]:
world = { "afghanistan":30.55,
"albania":2.77,
"algeria":39.21 }

for key, value in world.items():
    print(key + " -- " + str(value))

afghanistan -- 30.55
albania -- 2.77
algeria -- 39.21


## Numpy Array: Loop Over

In [39]:
import numpy as np
np_height = np.array([1.73, 1.68, 1.71, 1.89, 1.79])
np_weight = np.array([65.4, 59.2, 63.6, 88.4, 68.7])
bmi = np_weight / np_height ** 2
bmi 

array([21.85171573, 20.97505669, 21.75028214, 24.7473475 , 21.44127836])

* one-dimensional numpy array

In [40]:
type(bmi)

numpy.ndarray

In [41]:
for val in bmi :
    print(val)

21.85171572722109
20.97505668934241
21.750282138093777
24.74734749867025
21.44127836209856


## Two-Dimensional Numpy Array: Loop Over
* Same loop over for 1-D numpy array won't work
* To visit every element of an array, use nditer() function

In [48]:
import numpy as np
np_height = np.array([1.73, 1.68, 1.71, 1.89, 1.79])
np_weight = np.array([65.4, 59.2, 63.6, 88.4, 68.7])
two_dimensional_numpyarray = np.array([np_height, np_weight])

In [49]:
two_dimensional_numpyarray

array([[ 1.73,  1.68,  1.71,  1.89,  1.79],
       [65.4 , 59.2 , 63.6 , 88.4 , 68.7 ]])

* Visiting each list, but not each element

In [50]:
for val in two_dimensional_numpyarray:
    print(val)

[1.73 1.68 1.71 1.89 1.79]
[65.4 59.2 63.6 88.4 68.7]


* to iterate each element use np.nditer() function on 2d numpy array 
* printing all the height and weights element-wise

In [52]:
for val in np.nditer(two_dimensional_numpyarray):
    print(val)

1.73
1.68
1.71
1.89
1.79
65.4
59.2
63.6
88.4
68.7


### Exercises

Write a for loop that goes through each key:value pair of europe. On each iteration, "the capital of x is y" should be printed out, where x is the key and y is the value of the pair

In [55]:
# Definition of dictionary
europe = {'spain':'madrid', 'france':'paris', 'germany':'berlin',
          'norway':'oslo', 'italy':'rome', 'poland':'warsaw', 'austria':'vienna', 'bangladesh':'dhaka'}
          
# Iterate over europe
for country, capital in europe.items():
    print("the capital of", country, "is", capital)

the capital of spain is madrid
the capital of france is paris
the capital of germany is berlin
the capital of norway is oslo
the capital of italy is rome
the capital of poland is warsaw
the capital of austria is vienna
the capital of bangladesh is dhaka


## Pandas DataFrame: Loop Over
* loop over df means I have all rows available to be used

In [60]:
import pandas as pd 
brics = pd.read_csv('brics.csv', index_col=0)
brics

country    capital    area  population
BR        Brazil   Brasilia   8.516      200.40
RU        Russia     Moscow  17.100      143.50
IN         India  New Delhi   3.286     1252.00
CH         China    Beijing   9.597     1357.00
SA  South Africa   Pretoria   1.221       52.98

In [61]:
type(brics)

pandas.core.frame.DataFrame

* if a pandas dataframe were to function the same way as a 2D Numpy array, may be a basic for loop like this, to print out each row could work

* this prints only column names of DataFrame, this is not exactly we want

In [62]:
for val in brics:
    print(val)

country
capital
area
population


* In Pandas, mention explicit that you want to iterate over the rows 
* **iterrows() --> iterate each rows in a dataframe**
* use iterrows() method on the DF
* iterrows() method looks at the df, and on each iteration **generates two pieces of data: the label of row and secondly actual data in the row as a Pandas Series**
* store row label as lab, and row data as row. Thus, lab and row are arbitrary, any names can be given

In [66]:
brics 

country    capital    area  population
BR        Brazil   Brasilia   8.516      200.40
RU        Russia     Moscow  17.100      143.50
IN         India  New Delhi   3.286     1252.00
CH         China    Beijing   9.597     1357.00
SA  South Africa   Pretoria   1.221       52.98

* Because this row variable on each iteration is a Series, you can easily select additional information from it using subsetting techniques.

In [67]:
for lab, row in brics.iterrows():
    print(lab)
    print(row)

BR
country         Brazil
capital       Brasilia
area             8.516
population       200.4
Name: BR, dtype: object
RU
country       Russia
capital       Moscow
area            17.1
population     143.5
Name: RU, dtype: object
IN
country           India
capital       New Delhi
area              3.286
population         1252
Name: IN, dtype: object
CH
country         China
capital       Beijing
area            9.597
population       1357
Name: CH, dtype: object
SA
country       South Africa
capital           Pretoria
area                 1.221
population           52.98
Name: SA, dtype: object


* Suppose you only want to print out the capital in each iteration
* change print statement as follows:

In [68]:
for lab, row in brics.iterrows():
    print(lab + ": " + row['capital'])

BR: Brasilia
RU: Moscow
IN: New Delhi
CH: Beijing
SA: Pretoria


In [71]:
for lab, row in brics.iterrows():
    print(lab, row['area'])

BR 8.516
RU 17.1
IN 3.286
CH 9.597000000000001
SA 1.2209999999999999


In [74]:
for lab, row in brics.iterrows():
    print(lab, row['capital'], row['population'])

BR Brasilia 200.4
RU Moscow 143.5
IN New Delhi 1252.0
CH Beijing 1357.0
SA Pretoria 52.98


## Add column: DF

* Not efficient, because you are creating a Series object on every iteration

In [78]:
# All rows are available to us when we loop over df
for lab, row in brics.iterrows():
    brics.loc[lab, "name_length"] = len(row['country'])
print(brics)

         country    capital    area  population  name_length
BR        Brazil   Brasilia   8.516      200.40          6.0
RU        Russia     Moscow  17.100      143.50          6.0
IN         India  New Delhi   3.286     1252.00          5.0
CH         China    Beijing   9.597     1357.00          5.0
SA  South Africa   Pretoria   1.221       52.98         12.0


## apply() function
* use for loop, creates a Pandas Series object on every iteration. Thus, it is inefficient process.
* Way better approach, if you want to calculate an entire DataFrame column by applying a function on a particular column in an element-wise fashion. 
* do vectorized operation with apply() function
* **then use apply()**

## use of len function on a specific df column by apply function 
* when using len() function inside the apply function, no need to specify parenthesis or first bracket ()

In [80]:
brics["name_length_apply"] = brics['country'].apply(len)

## use of upper function on a specific df column by apply function 

In [93]:
brics["name_length_apply_2"] = brics['country'].apply(str.upper)

### Exercises

In [96]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Adapt for loop
for lab, row in cars.iterrows() :
    print(lab + ": " + str(row['cars_per_cap']))

US: 809
AUS: 731
JAP: 588
IN: 18
RU: 200
MOR: 70
EG: 45


In [97]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Code for loop that adds COUNTRY column
for lab, row in cars.iterrows():
    cars.loc[lab, "COUNTRY"] = row["country"].upper() 


# Print cars
print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT


* Replace the for loop with a one-liner that uses .apply(str.upper)

In [99]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Use .apply(str.upper)
cars['COUNTRY'] = cars['country'].apply(str.upper)
cars

cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT